### PR TITLE
[babel 8] Use ESM-based `chalk@5`

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

I updated `yarn-plugin-conditions` to allow defining ESM dependencies, so that we can start updating other deps for Babel 8.

~~Would this be ok for you other maintainers? It means that you can only run with `BABEL_8_BREAKING=true` if you are developing in the monorepo in ESM mode (it's enough to run `make use-esm` once).~~

~~The commits marked as `[15816]` are from #15816, please review that PR first.~~

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15792"></a>

### CODE REVIEW

**Review by JLHwung — APPROVED** [review]

**Review by liuxingbaoyu — APPROVED** [review]
To be honest, because of these special handling of chalk@5, I would rather stay with the old version unless there is a security problem.🤦‍♂️

File: packages/babel-standalone/test/helpers/env.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  // We only bundle Chalk 5 in local dev and in Babel 8, so we avoid this
+  // "polyfill" when releasing Babel 7 to make sure that we do not accidentally
+  // bundle Chalk 5.
+
+  if (process.env.BABEL_8_BREAKING) {
+    globalThis.navigator = {};
+  } else if (!process.env.IS_PUBLISH) {
+    global.navigator = {};
+  }
+}
```
Comment (liuxingbaoyu) [inline packages/babel-standalone/test/helpers/env.js:14]: Can we replace it with `navigator || {}` in rollup?
I remember someone seems to be using `@babel/standalone` in node, maybe it's not worth breaking them.
If you agree I can try to do this. :)

Comment (nicolo-ribaudo) [reply packages/babel-standalone/test/helpers/env.js:14]: Yes please!

**Comment by nicolo-ribaudo** [comment]
> To be honest, because of these special handling of chalk@5, I would rather stay with the old version unless there is a security problem.🤦‍♂️

The complexity is only there because we are using both chalk 2 and chalk 5 at the same time, if it was only chalk 5 it would be as complex as any other version. I would rather not have Babel 7 holding back Babel 8 😬 

The main advantages of using Chalk 5 are that:
- when using ESM, `yarn-plugin-conditions` works much better and thus it can properly load the TS definitions. Additionally, it exports the `Instance` type that it needs: this makes it possible for us to avoid this ugly import form: https://github.com/babel/babel/blob/78e504126d5d379ce1533e2324602f7a596e3e1d/packages/babel-code-frame/src/index.ts#L3-L8
- it's [much smaller](https://bundlephobia.com/package/chalk@5.3.0) than chalk 4 (75% smaller)

**Comment by liuxingbaoyu** [comment]
> it's [much smaller](https://bundlephobia.com/package/chalk@5.3.0) than chalk 4 (75% smaller)

Amazing! Now i like it! :)

**Comment by nicolo-ribaudo** [comment]
Good :)

Do you want to do https://github.com/babel/babel/pull/15792#discussion_r1305172580 here on in a separate PR?

**Comment by liuxingbaoyu** [comment]
It didn't look complicated, so I pushed it directly in this PR. :)

**Comment by liuxingbaoyu** [comment]
I ran into a little strange thing, even if I just build babel7, this code is still there.🤔
2{ RGC%TBH0OSS6__R_%Z4

**Comment by nicolo-ribaudo** [comment]
Unless you specify `STRIP_BABEL_8_BREAKNG`, the bundle will include both Babel 7 and Babel 8 code. Thus, I would also prefer to avoid the `BABEL_8_BREAKING` guard in the rollup config ans always replace `navigator`.

**Comment by nicolo-ribaudo** [comment]
@liuxingbaoyu I pushed a commit to use a Babel plugin instead. Rollup's replace plugin had some problems, because:
- by default, it does not replace things in member expressions
- it was also replacing `navigator` in a JSON file one of our dependencies was importing

**Comment by liuxingbaoyu** [comment]
> it was also replacing navigator in a JSON file one of our dependencies was importing

Yes, I found this problem too.
Thanks!

**Comment by liuxingbaoyu** [comment]
https://github.com/chalk/chalk/issues/598#issuecomment-1662499760
I think there is still a problem.
Currently its color detection is completely broken in non-browsers for this reason.🤦‍♂️

**Review by liuxingbaoyu — APPROVED** [review]
Sorry, I made a mistake. 😅
I thought `code-frame` and `highlight` would be affected too, now it looks fine.

**Comment by nicolo-ribaudo** [comment]
Yeah, just remember that (both in Babel 7 and Babel 8), you likely need to pass `forceColor: true` to `@babel/code-frame` if you want colors.

**Pull request merged**